### DISCUSSION TITLE
How to correctly infer the `options` type of `useQuery`?

### DISCUSSION THREAD

**Discussion by nkalpak** [trpc/trpc#3287] [Q&A]
To set the stage: I would like to avoid sprinkling `trpc.something.useQuery` calls inside the codebase, because this makes it difficult to setup some common options across API resources (for example, `trpc.note.getOne` needs to always have `staleTime: Infinity`).

The way I've historically done this with React Query is by using custom hooks which utilize a helper type to infer the options of queries and mutations through the query/mutation function (ex. `function useNote(options: InferQueryOptions<typeof getNoteApi>) { ... }`). Now that I'm using TRPC, I would like to do the same thing, and I've managed to get it working for [mutations](https://codesandbox.io/s/nervous-sanderson-m9vuvw?file=/src/mutation.ts).

The issue that I ran into is queries. The same approach which does work for mutations [doesn't seem to work for queries](https://codesandbox.io/s/nervous-sanderson-m9vuvw?file=/src/query.ts). So my question is really about how to approach this, what would be the idiomatic way of inferring query options for a particular procedure?

> I should note that I know that by doing this inference on the query itself I lose the type information for (ex.) `select`, and that if there is a `select` inside of the custom hook the return type of the custom hook will not reflect the return type of the `select`. I am OK with this limitation and I don't expect the TRPC solution to support this use-case.

**Comment by KATT** [comment]
I've seen someone doing this, but my googlefu is failing me. 

We should def add a helper for this, added https://github.com/trpc/trpc/issues/3293

**Reply by nkalpak** [reply]
I've managed to get something that works for my use case, sharing it [here](https://codesandbox.io/s/serene-ardinghelli-3tgesg?file=/src/utils.ts):
- [Query](https://codesandbox.io/s/serene-ardinghelli-3tgesg?file=/src/query.ts)
- [Mutation](https://codesandbox.io/s/serene-ardinghelli-3tgesg?file=/src/query.ts)

**Reply by KATT** [reply]
Wanna make a PR? :)

**Reply by nkalpak** [reply]
I'd love to! I assume this code would go inside `@trpc/react-query`, but where exactly in that sub-repo?

**Reply by KATT** [reply]
As long as it's exposed in index I don't care much :)

**Comment by transcendr** [comment]
Out of curiosity, re your full implementation, in your custom hook, are you merging the hook's own options (ie, `staleTime`) with the ones you're passing in?

**Reply by nkalpak** [reply]
> Out of curiosity, re your full implementation, in your custom hook, are you merging the hook's own options (ie, `staleTime`) with the ones you're passing in?

Yes that's the idea, though this doesn't work in the linked CodeSandbox as the query key is just an empty string there, while the TRPC wrapper hook expects it to be typed. The fully working implementation which doesn't have this issue can be seen [here](https://github.com/trpc/trpc/pull/3298/files#diff-c7f81882c47fa1ba6d264ac76d0987a8389d113a51a8517cff86cc59fe5d5dbbR12)
